### PR TITLE
feat: Add support for PSC network connections

### PR DESCRIPTION
**Pull request by hessjcg**

### CODE REVIEW

File: .github/renovate.json [hunk 1]
```diff
@@ -27,7 +27,7 @@
     },
     {
       "packageNames": ["com.google.apis:google-api-services-sqladmin"],
-      "allowedVersions": "/v1beta4-.*/"
+      "allowedVersions": "/v1-.*/"
```
Comment (enocom) [inline .github/renovate.json:30]: Let's leave v1beta4 in place. We do this for historical reasons. In short, v1beta4 should always be compatible with v1. To ensure that's the case, we continue using v1beta4 here.

**Review by enocom — CHANGES_REQUESTED** [review]
Let's stick to v1beta4. Otherwise this looks great.

File: core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java [hunk 2]
```diff
@@ -39,6 +39,8 @@ public class SqlAdminApiFetcherTest {
 
   public static final String SAMPLE_PUBLIC_IP = "34.1.2.3";
   public static final String SAMPLE_PRIVATE_IP = "10.0.0.1";
+  public static final String SAMPLE_PSC_IP = "10.0.0.2";
+  public static final String SAMPLE_PSC_HOSTNAME = "abcde.12345.us-central1.sql.goog";
```
Comment (jackwotherspoon) [inline core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java:43]: thoughts on maybe re-naming `SAMPLE_PSC_HOSTNAME` to `SAMPLE_PSC_DNSNAME` to be a bit more accurate?

File: docs/jdbc-postgres.md [hunk 3]
```diff
@@ -41,8 +41,14 @@ jdbc:postgresql:///<DATABASE_NAME>?cloudSqlInstance=<INSTANCE_CONNECTION_NAME>&s
...
 ### Specifying IP Types
- 
-"The `ipTypes` argument is used to specify a preferred order of IP types used to connect via a comma delimited list. For example, `ipTypes=PUBLIC,PRIVATE` will use the instance's Public IP if it exists, otherwise private. The value `ipTypes=PRIVATE` will force the Cloud SQL instance to connect via it's private IP. If not specified, the default used is `ipTypes=PUBLIC,PRIVATE`. 
+
+"The `ipTypes` argument is used to specify a preferred order of IP types used
+to connect via a comma delimited list. For example, `ipTypes=PUBLIC,PRIVATE`
+will use the instance's Public IP if it exists, otherwise private. The
+value `ipTypes=PRIVATE` will force the Cloud SQL instance to connect via
+it's private IP. The value `ipTypes=PSC` will force the Cloud SQL instance to
+connect to the database via [Private Service Connect](https://cloud.google.com/vpc/docs/private-service-connect).
```
Comment (jackwotherspoon) [inline docs/jdbc-postgres.md:50]:
```suggestion
it's private IP. The value `ipTypes=PSC` will force the the connection to the Cloud SQL instance via [Private Service Connect](https://cloud.google.com/vpc/docs/private-service-connect).
```

File: core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java (hunk 2 above)
Comment (hessjcg) [reply core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java:43]: Changed to SAMPLE_PSC_DNS_NAME.

File: .github/renovate.json (hunk 1 above)
Comment (hessjcg) [reply .github/renovate.json]: I switched back to v1beta4

File: core/src/test/java/com/google/cloud/sql/core/MockAdminApi.java [hunk 4]
```diff
@@ -65,11 +65,10 @@
 public class MockAdminApi {
 
   private static final Pattern CONNECT_SETTINGS_PATTERN =
-      Pattern.compile(
-          ".*/sql/v1beta4/projects/(?<project>.*)/instances/(?<instance>.*)/connectSettings");
+      Pattern.compile(".*/v1/projects/(?<project>.*)/instances/(?<instance>.*)/connectSettings");
   private static final Pattern GENERATE_EPHEMERAL_CERT_PATTERN =
       Pattern.compile(
-          ".*/sql/v1beta4/projects/(?<project>.*)/instances/(?<instance>.*):generateEphemeralCert");
+          ".*/v1/projects/(?<project>.*)/instances/(?<instance>.*):generateEphemeralCert");
```
Comment (jackwotherspoon) [inline core/src/test/java/com/google/cloud/sql/core/MockAdminApi.java:71]: should these stay as v1beta4 also?

File: core/src/main/java/com/google/cloud/sql/core/SqlAdminApiFetcher.java [hunk 5]
```diff
@@ -207,6 +207,11 @@ private Metadata fetchMetadata(CloudSqlInstanceName instanceName, AuthType authT
         ipAddrs.put(addr.getType(), addr.getIpAddress());
       }
 
+      // resolve DnsName into IP address for PSC
+      if (instanceMetadata.getDnsName() != null && !instanceMetadata.getDnsName().isEmpty()) {
+        ipAddrs.put("PSC", instanceMetadata.getDnsName());
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/SqlAdminApiFetcher.java:212]: nit: Making "PSC" a constant would be nice.

**Review by enocom — APPROVED** [review]

**Comment by hessjcg** [comment]
This was successfully hand-tested. The java connector successfully connected to a PSC database when `ipTypes=PSC` was set.

**Review by enocom — APPROVED** [review]

File: core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java [hunk 6]
```diff
@@ -63,6 +65,35 @@ public void testFetchInstanceData_returnsIpAddresses()
     Map<String, String> ipAddrs = instanceData.getIpAddrs();
     assertThat(ipAddrs.get("PRIMARY")).isEqualTo(SAMPLE_PUBLIC_IP);
     assertThat(ipAddrs.get("PRIVATE")).isEqualTo(SAMPLE_PRIVATE_IP);
+    assertThat(ipAddrs.get("PSC")).isEqualTo(SAMPLE_PCS_DNS_NAME);
+  }
+
+  @Test
+  public void testFetchInstanceData_returnsPscForNonIpDatabase()
```
Comment (hessjcg) [inline core/src/test/java/com/google/cloud/sql/core/SqlAdminApiFetcherTest.java:72]: This is the new test for PSC databases that have no IP address

**Review by jackwotherspoon — APPROVED** [review]
LGTM

**Pull request merged**